src/filters/token_router.rs

```diff
@@ -65,39 +62,39 @@ impl StaticFilter for TokenRouter {
 }
 
 impl Filter for TokenRouter {
-    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, ctx)))]
-    fn read(&self, mut ctx: ReadContext) -> Option<ReadResponse> {
+    fn read(&self, ctx: ReadContext) -> Option<ReadResponse> {
         match ctx.metadata.get(self.metadata_key.as_ref()) {
             None => {
-                if self.metrics.packets_dropped_no_token_found.get() % LOG_SAMPLING_RATE == 0 {
-                    error!(
-                        count = ?self.metrics.packets_dropped_no_token_found.get(),
-                        metadata_key = ?self.metadata_key.clone(),
-                        "Packets are being dropped as no routing token was found in filter dynamic metadata"
-                    );
-                }
+                tracing::trace!(
+                    metadata_key = %self.metadata_key,
+                    "Dropping packet, no routing token was found"
+                );
                 self.metrics.packets_dropped_no_token_found.inc();
                 None
             }
             Some(value) => match value {
                 Value::Bytes(token) => match ctx
                     .endpoints
-                    .retain(|e| e.metadata.known.tokens.contains(&**token))
+                    .iter()
+                    .find(|e| e.metadata.known.tokens.contains(&**token))
                 {
-                    RetainedItems::None => {
+                    Some(endpoint) => {
+                        tracing::trace!(%endpoint.address, token = &*base64::encode(token), "Endpoint matched");
+                        Some(ctx.into())
```

Comment:
I think this is a change in behaiviour - if I read this right, this will return the first `Endpoint` that matches the token?

Whereas previously it would return all `Endpoints` that had a token that matched.

Is that correct?